### PR TITLE
corretto{11,17,19}: init at 11.0.20.9.1/17.0.8.8.1/19.0.2.7.1

### LINKED ISSUE #244456: Package request: amazon corretto
**Project description**

Amazon Coretto JDK

**Metadata**

* homepage URL: https://aws.amazon.com/corretto/?filtered-posts.sort-by=item.additionalFields.createdDate&filtered-posts.sort-order=desc
* source URL: https://github.com/corretto
* license: GPL 2.0
* platforms: linux (and all else if possible)

### PR DESCRIPTION
**Pull request by rollf**
## Description of changes

This PR adds several versions of [Amazon Corretto](https://aws.amazon.com/corretto/). Corretto is basically OpenJDK with patches provided by Amazon. The Corretto derivations (in this PR) are based on the OpenJDK derivations (`overrideAttr`). I believe this is justified as a rebuild from scratch (`stdenv.mkDerivation`) would have to include all nixpkgs-specific fixes for OpenJDK, too.

There has been a [previous attempt](https://github.com/NixOS/nixpkgs/pull/55156) to include Corretto which was declined. I assume because it was stale? This PR would close https://github.com/NixOS/nixpkgs/issues/244456. Furthermore, this issue https://github.com/NixOS/nixpkgs/issues/114795 raises some concerns:
* I think the trademark is not a problem. The derivations make it clear, that this is *not* OpenJDK.
* Also, here, the distribution is built from source. No binary is patched. I therefore assume the running-modified-binaries argument is invalid.
* My personal use case for using Corretto is because my employer happens to use it. Our build environment even [enforces](https://docs.gradle.org/current/userguide/toolchains.html#sec:vendors) the vendor. I currently use OpenJDK on my NixOs machine and can't run an unmodified version of the build pipeline.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md); although I have some open questions:
  - See PR comments for details.

### CODE REVIEW

**Comment by rollf** [comment]
@edwtjo - you are the OpenJDK maintainer. Maybe you could have a look into this?

**Comment by rollf** [comment]
I haven't had any feedback for now and was looking into the commit history of the openjdk / adoption folders and came up with you. Maybe you can have a look into this?

@fabianhjr @Artturin @trofi @thiagokokada

**Comment by trofi** [comment]
I have no java usage or building experience. I can't provide the review above superficial styling bits.

I suggest posting your PR at https://discourse.nixos.org/t/prs-ready-for-review/3032 to get more eyes on it.

File: pkgs/development/compilers/corretto/11.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    jdk = jdk11;
+    gradle = gradle_7;
+    correttoVersion = "11.0.20.9.1";
+    src = fetchFromGitHub {
+      owner = "corretto";
+      repo = "corretto-11";
+      rev = "b880bdc8397ec3dd6b7cd4b837ce846c9e902783";
+      sha256 = "sha256-IomJHQn0ZgqsBZ5BrRqVrEOhTq7wjLiIKMQlz53JxsU=";
+    };
+    self = corretto11;
```
[the remark this answers is not in the document]
Comment (Artturin) [reply pkgs/development/compilers/corretto/11.nix:24]: self is used in python to do the same thing, but what you're doing here is much simpler so you can just use `finalAttrs.finalPackage`

https://ryantm.github.io/nixpkgs/using/overrides/#sec-pkg-overrideAttrs

File: pkgs/development/compilers/corretto/mk-corretto.nix [hunk 2]
```diff
@@ -0,0 +1,127 @@
+{ jdk
+, correttoVersion
```
[the remark this answers is not in the document]
Comment (Artturin) [reply pkgs/development/compilers/corretto/mk-corretto.nix:2]: yeah

File: pkgs/top-level/java-packages.nix [hunk 3]
```diff
@@ -93,6 +93,10 @@ in {
       ../development/compilers/adoptopenjdk-bin/jdk17-linux.nix
       ../development/compilers/adoptopenjdk-bin/jdk17-darwin.nix;
 
+    corretto11 = callPackage ../development/compilers/corretto/11.nix { };
+    corretto17 = callPackage ../development/compilers/corretto/17.nix { };
+    corretto19 = callPackage ../development/compilers/corretto/19.nix { };
```
[the remark this answers is not in the document]
Comment (Artturin) [reply pkgs/top-level/java-packages.nix:98]: I'ts fine as is

**Comment by rollf** [comment]
@Artturin Thank you for your review. I addressed your comments.

File: pkgs/development/compilers/corretto/mk-corretto.nix [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+  postPatch = ''
+    # The rpm/deb task definitions require a Gradle plugin which we don't
+    # have and so the build fails. We'll simply remove them here because
+    # they are not needed anyways.
+    rm -rf installers/linux/universal/{rpm,deb}
+
+    # `/usr/bin/rsync` is invoked to copy the source tree. We don't have that.
+    for file in $(find installers -name "build.gradle"); do
+      substituteInPlace $file --replace "workingDir '/usr/bin'" "workingDir '.'"
```
Comment (Artturin) [inline pkgs/development/compilers/corretto/mk-corretto.nix:41]: Is rsync needed in `nativeBuildInputs`

**Comment by Artturin** [comment]
Update the commit title to match the pr title

File: pkgs/development/compilers/corretto/mk-corretto.nix (hunk 4 above)
Comment (rollf) [reply pkgs/development/compilers/corretto/mk-corretto.nix:41]: I'm note sure what the question is. `rsync` is invoked during the build (see [here)](https://github.com/corretto/corretto-11/blob/develop/installers/linux/universal/tar/build.gradle#L81). I have opted to keep that but simply make sure that (the nixified) `rsync` can be invoked.

Comment (Artturin) [reply pkgs/development/compilers/corretto/mk-corretto.nix:41]: Ahh I didn't know the context so it was weird that the replacement didn't mention rsync

**Comment by rollf** [comment]
@Artturin Thanks for your feedback and merge!

**Pull request merged**